### PR TITLE
feat: generate CI definition from selected pipeline (graph)

### PR DESCRIPTION
**Pull request by dnitsch**
Lays down groundwork for #16 
and adds initial support first pass for #16  with `github` type implementation see [here](https://github.com/Ensono/taskctl/actions/runs/11739797021) an example auto generated pipeline running in GHA

> This merge will mark the bump to v2 and will keep everything as pre-release until further work to ensure stability. 

Notable work:
- add sort able methods on the StageNode
- add graph denormalization methods to enable full path ancestor tracing and hence allow the ability to inherit variables and other metadata in a cascading way.
- fix up graph command to show full dependency graph in a normalized way (+ add mermaid support in graph command)

Docs:
- [overview of graph changes](https://github.com/Ensono/taskctl/blob/feat/gen-ci-def-v0/docs/graph-implementation.md)
- [overview of generate v0 implementation](https://github.com/Ensono/taskctl/blob/feat/gen-ci-def-v0/docs/ci-generator.md)

### CODE REVIEW

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -1,5 +1,6 @@
 name: Release
 
+###_BEGIN_TASKCTL_SKIP_###
```
Comment (ElvenSpellmaker) [inline .github/workflows/release.yml:3]: Does this still do anything?

File: cmd/taskctl/generate.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+		Args:         cobra.MinimumNArgs(0),
+		SilenceUsage: true,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			conf, err := rootCmd.initConfig()
+			if err != nil {
+				return err
+			}
+			// if len(f.targetTyp) == 0 {
+			// 	return fmt.Errorf("target ")
+			// }
```
Comment (ElvenSpellmaker) [inline cmd/taskctl/generate.go:38]: Commented out?

File: cmd/taskctl/run.go [hunk 3]
```diff
@@ -119,20 +149,32 @@ func (r *runCmd) runTarget(taskRunner *runner.TaskRunner, conf *config.Config, a
 func (r *runCmd) runPipeline(g *scheduler.ExecutionGraph, taskRunner *runner.TaskRunner, summary bool) error {
 	sd := scheduler.NewScheduler(taskRunner)
 	go func() {
-		<-cancel
+		<-cancel // r.ctx.Done()
```
Comment (ElvenSpellmaker) [inline cmd/taskctl/run.go:152]: Comment?

File: cmd/taskctl/taskctl.go [hunk 4]
```diff
@@ -225,7 +212,7 @@ func (tc *TaskCtlCmd) buildTaskRunner(args []string, conf *config.Config) (*runn
 	}
 
 	go func() {
-		<-cancel
+		<-cancel // tc.Cmd.Context().Done()
```
Comment (ElvenSpellmaker) [inline cmd/taskctl/taskctl.go:215]: Comment?

File: cmd/taskctl/watch.go [hunk 5]
```diff
@@ -30,7 +30,7 @@ func newWatchCmd(rootCmd *TaskCtlCmd) {
 				wg.Add(1)
 
 				go func(w *watch.Watcher) {
-					<-cancel
+					<-cancel //rootCmd.Cmd.Context().Done()
```
Comment (ElvenSpellmaker) [inline cmd/taskctl/watch.go:33]: Comment?

File: internal/genci/genci.go [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+	"github.com/Ensono/taskctl/pkg/scheduler"
+	"gopkg.in/yaml.v2"
+)
+
+var (
+	ErrImplementationNotExist   = errors.New("implementation does not exist")
+	ErrFailedImplementationInit = errors.New("failed to initialise the implementation")
+)
+
+// const taskctlVersion string = "v1.8."
```
Comment (ElvenSpellmaker) [inline internal/genci/genci.go:29]: hmm?

File: internal/genci/genci.go [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+	switch implTyp {
+	case GitHubCITarget:
+		gh, err := newGithubCiImpl(conf)
+		if err != nil {
+			return nil, fmt.Errorf("%w, %v", ErrFailedImplementationInit, err)
+		}
+		gci.implementation = gh
+	// case GitlabCITarget:
+	// 	gci.implementation = &DefualtCiImpl{}
+	// TODO: add more here
```
Comment (ElvenSpellmaker) [inline internal/genci/genci.go:72]: Huh and also TODO?

File: internal/genci/genci.go [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+}
+
+func extractGeneratorMetadata[T any](implTyp CITarget, generatorMeta map[string]any) (*T, error) {
+	typ := new(T)
+	if gh, found := generatorMeta[string(implTyp)]; found {
+		b, err := yaml.Marshal(gh)
+		if err != nil {
+			return typ, err
+		}
+		if err := yaml.Unmarshal(b, typ); err != nil {
```
Comment (ElvenSpellmaker) [inline internal/genci/genci.go:96]: Why do you marshal and then unmarshal here?

File: internal/genci/githubimpl.go [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+	"github.com/Ensono/taskctl/pkg/scheduler"
+	"gopkg.in/yaml.v2"
+)
+
+var ErrInvalidCiMeta = errors.New("CI meta is invalid")
+
+// githubCiImpl is the implementation of GHA pipeline generation from TaskCtl ExecutionGraph
+// The graph has to be denormalized to ensure that all env variables are correctly cascaded to the tasks
+type githubCiImpl struct {
+	// TODO: internal props
```
Comment (ElvenSpellmaker) [inline internal/genci/githubimpl.go:21]: Yo, internal props?

File: internal/schema/github.go [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+
+	"gopkg.in/yaml.v2"
+)
+
+// Workflow represents the root of a GitHub workflow file.
+type GithubWorkflow struct {
+	Name string               `json:"name,omitempty" yaml:"name,omitempty"`
+	On   *GithubTriggerEvents `json:"on" yaml:"on"`
+	Jobs yaml.MapSlice        `json:"jobs" yaml:"jobs"`
+	// Jobs        yaml.Node         `json:"jobs" yaml:"jobs"`
```
Comment (ElvenSpellmaker) [inline internal/schema/github.go:15]: Yeah?

File: pkg/output/output.go [hunk 11]
```diff
@@ -60,12 +60,14 @@ func (t *TaskOutput) WithCloseCh(closeCh chan bool) *TaskOutput {
 
 // Stdout returns io.Writer that can be used for Job's STDOUT
 func (o *TaskOutput) Stdout() io.Writer {
-	return MultiWriter(o.decorator, o.t.Log.Stdout)
+	// return MultiWriter(o.decorator, o.t.Log.Stdout)
```
Comment (ElvenSpellmaker) [inline pkg/output/output.go:63]: comment?

File: pkg/output/output.go [hunk 12]
```diff
@@ -60,12 +60,14 @@ func (t *TaskOutput) WithCloseCh(closeCh chan bool) *TaskOutput {
...
+	return o.decorator
 }
 
 // Stderr returns io.Writer that can be used for Job's STDERR
 func (o *TaskOutput) Stderr() io.Writer {
-	return MultiWriter(o.decorator, o.t.Log.Stderr)
+	return o.decorator
+	// return MultiWriter(o.decorator, o.t.Log.Stderr)
```
Comment (ElvenSpellmaker) [inline pkg/output/output.go:70]: comment?

File: pkg/output/prefixed.go [hunk 13]
```diff
@@ -28,51 +21,29 @@ func NewPrefixedOutputWriter(t *task.Task, w io.Writer) *prefixedOutputDecorator
...
+			return 0, err
+		}
+		if advance == 0 {
+			break
 		}
-		// skip writing empty lines
+		// do not write empty lines
 		if len(line) == 0 {
-			continue
+			// p = p[advance:]
```
Comment (ElvenSpellmaker) [inline pkg/output/prefixed.go:38]: Sure?

File: pkg/runner/context.go [hunk 14]
```diff
@@ -45,33 +46,51 @@ type ExecutionContext struct {
...
+	c.down = down
+	c.before = before
+	c.after = after
+
 	return c
 }
 
+func WithContainerOpts(containerOpts *utils.Container) ExecutionContextOption {
+	return func(c *ExecutionContext) {
+		// TODO: make this smarter...
```
Comment (ElvenSpellmaker) [inline pkg/runner/context.go:85]: Is this a TODO for this PR?

File: pkg/runner/context.go [hunk 15]
```diff
@@ -146,8 +167,11 @@ func (c *ExecutionContext) GenerateEnvfile(env variables.Container) error {
 	// create a string builder object to hold all of the lines that need to be written out to
 	// the resultant file
 	builder := []string{}
-
+	// if envfile path was provided
+	// if
+	// utils.ReadEnvFile()
 	// iterate around all of the environment variables and add the selected ones to the builder
+	// env container at this point should already include all the merged variables by precedence
```
Comment (ElvenSpellmaker) [inline pkg/runner/context.go:174]: This comment reads a little odd

File: pkg/task/task.go [hunk 16]
```diff
@@ -52,23 +50,25 @@ type Task struct {
...
+	end      time.Time
+	skipped  bool
+	exitCode int16
+	errored  bool
+	errorVal error
+	// Log      struct {
+	// 	Stderr *bytes.Buffer
+	// 	Stdout *bytes.Buffer
+	// }
+	// Generator Task Level
```
Comment (ElvenSpellmaker) [inline pkg/task/task.go:70]: needed?

File: pkg/task/task.go [hunk 17]
```diff
@@ -77,17 +77,118 @@ func NewTask(name string) *Task {
...
+		exitCode:  -1,
+		errored:   false,
+		mu:        sync.Mutex{},
+		// Log: struct {
+		// 	Stderr *bytes.Buffer
+		// 	Stdout *bytes.Buffer
+		// }{
+		// 	Stderr: &bytes.Buffer{},
+		// 	Stdout: &bytes.Buffer{},
+		// },
```
Comment (ElvenSpellmaker) [inline pkg/task/task.go:89]: needed?

File: pkg/task/task.go [hunk 18]
```diff
@@ -77,17 +77,118 @@ func NewTask(name string) *Task {
...
+	t.AllowFailure = task.AllowFailure
+	t.After = task.After
+	t.Before = task.Before
+	t.Interactive = task.Interactive
+	t.ResetContext = task.ResetContext
+	t.Condition = task.Condition
+	t.Artifacts = task.Artifacts
+	t.Description = task.Description
+	t.Generator = task.Generator
+	// merge vars
```
Comment (ElvenSpellmaker) [inline pkg/task/task.go:108]: hmm?

File: pkg/task/task.go [hunk 19]
```diff
@@ -97,30 +198,30 @@ func FromCommands(name string, commands ...string) *Task {
...
 
-	if t.Log.Stderr.Len() > 0 {
-		return utils.LastLine(t.Log.Stderr)
-	}
+	// if t.Log.Stderr.Len() > 0 {
+	// 	return utils.LastLine(t.Log.Stderr)
+	// }
 
-	return utils.LastLine(t.Log.Stdout)
+	// return utils.LastLine(t.Error())
```
Comment (ElvenSpellmaker) [inline pkg/task/task.go:219]: needed?

File: pkg/variables/variables.go [hunk 20]
```diff
@@ -66,8 +61,13 @@ func (vars *Variables) Map() map[string]interface{} {
...
 }
 
-// Merge merges two Containers into new one
-func (vars *Variables) Merge(src Container) Container {
+// Merge merges into current container with the src Container
+// src will overwrite the existing keys if exists
+// returns a the merged Container
+//
+// TODO: since we are dealing with pointers this does not need to return anything
+// instead update only the underlying reference, create new story for this.
```
Comment (ElvenSpellmaker) [inline pkg/variables/variables.go:69]: Is this a TODO for this PR?

File: .github/workflows/release.yml (hunk 1 above)
Comment (dnitsch) [reply .github/workflows/release.yml:3]: not right now - probably will never be implemented, but would rather delete it all as part of the v2 stabilisation

File: cmd/taskctl/taskctl.go (hunk 4 above)
Comment (dnitsch) [reply cmd/taskctl/taskctl.go:215]: I would want to keep these in here for now as they need to be properly implemented

listening to os.Signals isn't done properly and we don't shut down gracefully

File: internal/genci/genci.go (hunk 8 above)
Comment (dnitsch) [reply internal/genci/genci.go:96]: there is no real safe way to force the `any` type into a byte slice which is what the Unmarshal method expects

File: pkg/variables/variables.go (hunk 20 above)
Comment (dnitsch) [reply pkg/variables/variables.go:69]: next PR - ideally no more changes in this one 🤔

File: pkg/variables/variables.go [hunk 21]
```diff
@@ -1,32 +1,27 @@
+// Package variables is a thin wrapper over the sync.Map package
+//
+// TODO: The sync.Map implementation may not really be necessary anymore since we denormalize the graph
+// any/all vars and env vars will have their own address space so collision _will_ not occur.
+// "¯\_(ツ)_/¯"
```
[the remark this answers is not in the document]
Comment (dnitsch) [reply pkg/variables/variables.go:5]: "¯\_(ツ)_/¯"

**Review by ElvenSpellmaker — APPROVED** [review]

**Pull request merged**